frontend/catalyst/jax_primitives.py

```diff
@@ -551,13 +559,87 @@ def _apply_registered_pass_lowering(
 
 
 #
-# func
+# module
 #
-mlir_fn_cache: Dict["catalyst.jax_tracer.Function", Any] = {}
+
+mlir_fn_cache: Dict[Any, Any] = {}
+
+
+@module_p.def_impl
+def _module_def_impl(*args, call_jaxpr, fn):  # pragma: no cover
+    raise NotImplementedError()
+
+
+def _module_to_mlir(ctx, *args, call_jaxpr, fn, call=True):
+
+    if not isinstance(fn, qml.QNode) and not call:
+        _func_lowering(ctx, *args, call_jaxpr=call_jaxpr, fn=fn, call=call)
```

Comment:
> Should the branching on QNode (our modules) be done outside this function?

I actually like the design of universally using `module_p` everywhere during tracing and dispatching to `module_p`/`func_p` depending on whether function is a qnode inside lowering. It means as a programmer when writing tracing I don't need to worry about "should I use module or func primitive" myself.

I think the reverse works too: using `func_p` universally during tracing (so no change required in tracing since its already `func_p`), but dispatch to `module_p` if the function is a qnode. I believe there already is [such a check](https://github.com/PennyLaneAI/catalyst/blob/main/frontend/catalyst/jax_primitives.py#L573) in `func_p`'s lowering, just that we can now do different things at that branching (instead of marking a `qnode` attribute we enter this new `module_p`). 

On second thought I think this approach is the best one, since I understand that to be the whole point of this work (don't use attribute, but nested modules, to indicate a qnode scope).

Reply:
resolved